src/cargo/ops/cargo_rustc/context.rs

```diff
@@ -478,9 +481,11 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
         // doing this eventually.
         let __cargo_default_lib_metadata = env::var("__CARGO_DEFAULT_LIB_METADATA");
         if !unit.profile.test &&
-            (unit.target.is_dylib() || unit.target.is_cdylib()) &&
+            (unit.target.is_dylib() || unit.target.is_cdylib() ||
+                 (unit.target.is_bin() && self.target_triple().starts_with("wasm32-"))) &&
```

Comment:
Could this be refactored into a dedicated method? Something like:

```rust
fn fails_to_work_with_mangled_filename(...) -> bool {
    // ...
}
```

(just to make it a bit more clear as to the intent)